### PR TITLE
remove maude backend and make java the default

### PR DESCRIPTION
**Pull request by dwightguth**
**please review**

### CODE REVIEW

**Comment by dwightguth** [comment]
**DO NOT MERGE**. This pull request needs to be merged as part of the release process for K 3.5, so that all changes up to this diff are part of K 3.5 and this diff is part of K 3.6.

@yilongli please review changes to java backend and java code

@grosu please review changes to documentation files and tutorial

@radumereuta please review changes to parser and prelude

@traiansf please review changes to agent

@yilongli @radumereuta please review everything else.

@cos you're welcome to comment as well if you like. You might be interested in seeing what I did to deal with productions that declare KLabel constants.

**Comment by radumereuta** [comment]
I'm about to board my flight to Europe, so the best I can do is Thursday.

**Comment by yilongli** [comment]
The changes to the java backend look fine to me.

**Comment by traiansf** [comment]
changes to agent look good to me.  Does quote/unquote work now with the java backend?

File: kernel/src/main/java/org/kframework/kil/Production.java [hunk 1]
```diff
@@ -215,6 +216,21 @@ public int getArity() {
         return arity;
     }
 
+    public int getArityOfKItem() {
```
Comment (cos) [inline kernel/src/main/java/org/kframework/kil/Production.java:238]: could you add some javadoc explaining the distinction between this method and `getArity`?

File: kernel/src/main/java/org/kframework/kil/Production.java [hunk 2]
```diff
@@ -391,4 +407,14 @@ public void setBinderMap(Multimap<Integer, Integer> binderMap) {
     public void setChildren(List<ProductionItem> children, Enum<?> _void) {
         this.items = children;
     }
+
+    public String getKLabelOfKItem() {
```
Comment (cos) [inline kernel/src/main/java/org/kframework/kil/Production.java:437]: and between this method and the simple `getKLabel`

**Comment by grosu** [comment]
Looks excellent to me!  I am so glad to see this pull request.

File: k-distribution/include/builtins/model-checker.k [hunk 3]
```diff
@@ -2,7 +2,6 @@
 require "bool.k"
 
 module LTL-SYNTAX-HOOKS
```
Comment (radumereuta) [inline k-distribution/include/builtins/model-checker.k:4]: Is this file still needed? I think this worked only with the maude backend.

File: k-distribution/include/io/uris.k [hunk 4]
```diff
@@ -1,94 +0,0 @@
...
-  rule #string2klist(K:K) => K when isString(K) =/=K true
-
-  syntax KList ::= #statTypes(KList) [function]
-  rule #statTypes(S1:String,,S2:String,,S3:String,,S4:String,,S5:String,,S6:String,,S7:String,,S8:String,,S9:String,,S10:String,,S11:String,,S12:String) => String2Int(S1),,String2Int(S2),,String2Int(S3),,String2Bool(S4),,String2Bool(S5),,String2Bool(S6),,String2Int(S7),,String2Int(S8),,String2Int(S9),,String2Int(S10),,String2Int(S11),,String2Int(S12)
-  rule #statTypes(K:K) => K
-
-// ------------------------------
-// Dynamic parsing
-
-  rule #parse(StringToParse:String, Sort:String) => 'String2K(#tcpRequest("parse\x01" +String Sort +String "\x01" +String StringToParse +String "\x01"))::K
```
Comment (radumereuta) [inline k-distribution/include/io/uris.k:78]: I haven't gotten to the end of the diff yet, but what happens to this?

File: k-distribution/include/builtins/model-checker.k (hunk 3 above)
Comment (dwightguth) [inline k-distribution/include/builtins/model-checker.k:4]: I would like to keep it; it defines the LTL API that Traian's work on model checking will have to support. We can make adjustments to that API of course, but it needs to stick around as a base, otherwise we'll have to redo everything from scratch.

File: k-distribution/include/io/uris.k (hunk 4 above)
Comment (dwightguth) [inline k-distribution/include/io/uris.k:78]: It still works; they are implemented as hooks in the java engine. It is only the janky tcp mechanism for communicating with maude that has been affected. There are a few hooks that were implemented in maude only, but these can be reimplemented as needed.

File: kernel/src/main/java/org/kframework/kil/Production.java [hunk 5]
```diff
@@ -169,13 +170,19 @@ public String getLabel() {
...
+     * KItem (ie, is a function or a constructor).
+     * @return
+     */
     public String getKLabel() {
         String klabel = getAttribute("klabel");
         if (klabel == null && isSyntacticSubsort()) {
             return null;
         } else if (klabel == null) {
             if (sort.equals(Sort.KLABEL) && getArity() == 0)
-                klabel = getPrefixLabel();
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/kil/Production.java:178]: I don't think `getPrefixLabel` should be integrated into `getKLabel`. That function was created separately only because of maude.

File: kernel/src/main/java/org/kframework/kil/Syntax.java [hunk 6]
```diff
@@ -90,8 +90,9 @@ public String toString() {
             return lbls;
         for (PriorityBlock pb : priorityBlocks) {
             for (Production prod : pb.getProductions()) {
-                if (prod.getSort().isComputationSort() || prod.getSort().equals(Sort.KLABEL) && prod.isConstant())
-                    lbls.add(prod.getKLabel());
+                if (prod.getSort().isComputationSort() || prod.getSort().equals(Sort.KLABEL) && prod.isConstant()) {
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/kil/Syntax.java:93]: `isConstant` in the future is going to be true iff it has the attribute `token`.
Is this condition still true with this information?

**Comment by radumereuta** [comment]
lgtm

File: kernel/src/main/java/org/kframework/kil/Production.java (hunk 5 above)
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/kil/Production.java:178]: Note I am actually removing a call to getPrefixLabel here. The issue is that `syntax KLabel ::= "foo"` doesn't declare a construct whose KLabel is "foo", it's declaring a KLabel /itself/, so returning "foo" from this method leads to incorrect assumptions in the code later on, such as that the arity of a production is always equal to the arity of KItems whose KLabel comes from the KLabel of that production. In any event, this is all temporary hackiness in the implementation. I believe that in kore, there should never be any productions whose return sort is "KLabel". KLabel declarations like the one above in full K should be syntactic sugar for " `syntax KItem ::= foo(KList)` or `syntax KItem ::= foo(K, K)` depending on whether an arity has been declared. And only KLabels with an arity can heat/cool.

File: kernel/src/main/java/org/kframework/kil/Syntax.java (hunk 6 above)
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/kil/Syntax.java:93]: See above: I believe this method should eventually go away.

File: kernel/src/main/java/org/kframework/kil/Production.java (hunk 5 above)
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/kil/Production.java:178]: @cos @grosu you are the kore gatekeepers, does this sound like an acceptable solution? In particular it's the invariant mentioned in the above post that I would like to guarantee for all KItems.

Comment (grosu) [inline kernel/src/main/java/org/kframework/kil/Production.java:178]: LGTM (and yes, this code will change soon, hopefully, because this is just a deconstructor for a KORE syntactic constructor; the plan is that the KORE library will provide implementations for all the constructors and the associated deconstructors).

Syntactically, KORE does not restrict you to declare `syntax KLabel ::= "foo"`.  But we are going to have all sorts of additional restrictions that will be checked and enforced after parsing.  Not sure yet whether this will be one of them or not, but we do not have to decide it now.

**Comment by dwightguth** [comment]
Jenkins: test this please

**Comment by dwightguth** [comment]
Jenkins: test this please
